src/jltypes.c

```diff
@@ -797,6 +788,26 @@ jl_value_t *jl_cache_type_(jl_datatype_t *type)
     return (jl_value_t*)type;
 }
 
+int jl_type_equality_is_identity(jl_value_t *t1, jl_value_t *t2)
+{
+    if (t1 == t2)
+        return 1;
+    if (!jl_is_datatype(t1) || !jl_is_datatype(t2))
+        return 0;
+    jl_datatype_t *dt1 = (jl_datatype_t *) t1;
+    jl_datatype_t *dt2 = (jl_datatype_t *) t2;
+    if (!(is_cacheable(dt1) || jl_svec_len(dt1->parameters) == 0)) {
```

Comment:
could be nice to have some comments here about what the reasoning here is, what the parameters are, etc.